src/pip/_internal/cli/cmdoptions.py

```diff
@@ -920,6 +922,35 @@ def check_list_path_option(options):
 )  # type: Callable[..., Option]
 
 
+discovered_auth_plugins = {
+    entry_point.name: entry_point
+    for entry_point
+    in iter_entry_points('pip.auth_plugins')
+}
+
+
+def _handle_auth_plugin(option, opt_str, value, parser):
+    # type: (Option, str, str, OptionParser) -> None
+    plugin = discovered_auth_plugins[value].load()
+    setattr(parser.values, option.dest, plugin.Auth)
+
+
+auth_plugin = partial(
+    PipOption,
+    '--auth', '--auth-plugin',
+    dest='auth_class',
+    type='choice',
+    metavar='plugin',
+    default=None,
+    choices=list(discovered_auth_plugins.keys()),
+    action='callback',
+    callback=_handle_auth_plugin,
+    help='Specify authentication plugin to be used: {}.'.format(
+        ', '.join(discovered_auth_plugins.keys()) or 'no plugins found'
```

Comment:
I wonder if we can phrase this better in case *no plugin found*.  Personally I'm bad at making concise sentences so I can't offer and suggestion though.